File: frontend/app/src/api/learningPath.ts
import request from '../utils/request'
import type { DbResource } from './resource'

export type PublicLearningPath = {
  id: number
  title: string
  type?: string | null
  description?: string | null
  is_public: boolean
  is_active: boolean
  cover_image_url?: string | null
  category_id?: number | null
  category_name?: string | null
}

export type LearningPathDisplayBase = {
  id: string
  title: string
  description: string
  thumbnail: string
  categoryName: string
}

export function mapPublicLearningPathToDisplayBase(p: PublicLearningPath): LearningPathDisplayBase {
  const id = String(p.id)
  const title = String(p.title || '').trim() || `Path ${id}`
  const description = String(p.description || '').trim()
  const thumbnail = String(p.cover_image_url || '').trim()
  const categoryName = String(p.category_name || '').trim()
  return {
    id,
    title,
    description,
    thumbnail,
    categoryName,
  }
}

export type PublicLearningPathDetail = PublicLearningPath & {
  path_items: Array<{
    id: number
    learning_path_id: number
    resource_id: number
    resource_type: string
    title: string
    order_index: number
    stage?: string | null
    purpose?: string | null
    estimated_time?: number | null
    is_optional: boolean
    resource_data?: DbResource | null
  }>
}

export function listPublicLearningPaths() {
  return request.get<PublicLearningPath[], PublicLearningPath[]>('/learning-paths/public')
}

export function getPublicLearningPathDetail(id: number) {
  return request.get<PublicLearningPathDetail, PublicLearningPathDetail>(`/learning-paths/public/${id}`)
}

export function createLearningPath(payload: {
  title: string
  type?: string | null
  description?: string
  is_public: boolean
  cover_image_url?: string | null
}) {
  return request.post<PublicLearningPath, PublicLearningPath>('/learning-paths/', payload)
}

export function createLearningPathWithCategory(payload: {
  title: string
  type?: string | null
  description?: string
  is_public: boolean
  cover_image_url?: string | null
  category_id?: number | null
}) {
  return request.post<PublicLearningPath, PublicLearningPath>('/learning-paths/', payload)
}

export type MyLearningPath = PublicLearningPath

export type MyLearningPathDetail = PublicLearningPathDetail

export function listMyLearningPaths() {
  return request.get<MyLearningPath[], MyLearningPath[]>('/learning-paths/')
}

export function getMyLearningPathDetail(id: number) {
  return request.get<MyLearningPathDetail, MyLearningPathDetail>(`/learning-paths/${id}`)
}

export function updateMyLearningPath(
  id: number,
  payload: {
    title?: string
    type?: string | null
    description?: string
    is_public?: boolean
    cover_image_url?: string | null
    category_id?: number | null
  },
) {
  return request.patch<MyLearningPath, MyLearningPath>(`/learning-paths/${id}`, payload)
}

export type AttachLearningPathResult = {
  already_exists: boolean
  learning_path: MyLearningPath
}

export function attachPublicLearningPathToMe(id: number) {
  return request.post<AttachLearningPathResult, AttachLearningPathResult>(`/learning-paths/me/${id}/attach`, {})
}

export function deleteMyLearningPath(id: number) {
  return request.delete(`/learning-paths/${id}`)
}

export function addResourceToMyLearningPath(
  learningPathId: number,
  payload: {
    resource_id: number
    order_index?: number
    stage?: string | null
    purpose?: string | null
    estimated_time?: number | null
    is_optional?: boolean
  },
) {
  return request.post(`/learning-paths/${learningPathId}/items`, payload)
}

export function removeResourceFromMyLearningPath(learningPathId: number, resourceId: number) {
  return request.delete(`/learning-paths/${learningPathId}/items/${resourceId}`)
}

File: backend/app/schemas/learning_path.py
from typing import Optional, List
from enum import Enum
from pydantic import BaseModel

from app.schemas.resources.resource import ResourceResponse


class ResourceKind(str, Enum):
    video = "video"
    document = "document"
    article = "article"


class LearningPathBase(BaseModel):
    title: str
    description: Optional[str] = None
    is_public: bool = False
    cover_image_url: Optional[str] = None
    category_id: int
    category_name: Optional[str] = None

    model_config = {
        "from_attributes": True
    }


class LearningPathCreate(LearningPathBase):
    # 与 CURD 保持一致，创建时主要使用 title/description
    pass


class LearningPathUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    is_public: Optional[bool] = None
    is_active: Optional[bool] = None
    cover_image_url: Optional[str] = None
    category_id: Optional[int] = None

    model_config = {
        "from_attributes": True
    }


class PathItemInLearningPathResponse(BaseModel):
    id: int
    learning_path_id: int
    resource_id: int
    resource_type: ResourceKind
    title: str
    order_index: int
    stage: Optional[str] = None
    purpose: Optional[str] = None
    estimated_time: Optional[int] = None
    is_optional: bool = False
    # 按需返回嵌入的资源详情
    resource_data: Optional[ResourceResponse] = None

    model_config = {
        "from_attributes": True
    }


class LearningPathResponse(LearningPathBase):
    id: int
    is_active: bool = True

    model_config = {
        "from_attributes": True
    }


class LearningPathDetailResponse(LearningPathResponse):
    path_items: List[PathItemInLearningPathResponse] = []

    model_config = {
        "from_attributes": True
    }


class AddResourceToLearningPathRequest(BaseModel):
    resource_id: int
    order_index: Optional[int] = None
    stage: Optional[str] = None
    purpose: Optional[str] = None
    estimated_time: Optional[int] = None
    is_optional: Optional[bool] = None

    model_config = {
        "from_attributes": True
    }


class LearningPathAttachResponse(BaseModel):
    already_exists: bool
    learning_path: LearningPathResponse

    model_config = {
        "from_attributes": True
    }

    